### PR TITLE
[#14649][feat] AutoDeploy: UltraV3 nvfp4 MTP

### LINKED ISSUE #14649: [Feature]: AutoDeploy: UltraV3 nvfp4 + MTP
### 🚀 The feature, motivation and pitch

Integration / accuracy test to make sure this works + necessary changes.

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the [documentation](https://nvidia.github.io/TensorRT-LLM/) and [examples](https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples) for answers to frequently asked questions.

### PR DESCRIPTION
**Pull request by govind-ramnarayan**
Fixes: #14649

## Summary by CodeRabbit

* **New Features**
  * Added support for Ultra V3 model with MTP speculative decoding for enhanced inference performance.
  * Introduced NVFP4 quantization with FP8 KV cache configuration options.
  * Extended model deployment configurations with advanced optimization parameters including multi-stream and fusion settings.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA/TensorRT-LLM/pull/14473?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 1]
```diff
@@ -373,7 +401,18 @@ def get_sharding_config(self) -> Dict[str, Any]:
     # TODO(govind): It's possible that draft models have different quant configs than target models.
     # We need to address this possibility.
     def get_quant_config(self) -> Dict[str, Any]:
-        return self.target_factory.get_quant_config()
+        qcfg = dict(self.target_factory.get_quant_config())
+        excluded = qcfg.get("exclude_modules")
+        if excluded:
+            conversion_mapping = (
+                self.draft_factory.get_exclude_modules_conversion_mapping()
+                or self.draft_factory.get_checkpoint_conversion_mapping()
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: nit: no reason to get checkpoint conversion mapping anymore. Just do exclude modules conversion mapping or nothing - and effectively identity map if nothing is passed.

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 2]
```diff
@@ -78,13 +97,22 @@ def _build_model(self, device: DeviceLikeType) -> nn.Module:
...
         self._checkpoint_conversion_mapping = getattr(model, "_checkpoint_conversion_mapping", None)
+        self._exclude_modules_conversion_mapping = getattr(
+            model, "_exclude_modules_conversion_mapping", None
+        )
 
         model.eval()
 
         return model
 
+    def get_checkpoint_conversion_mapping(self) -> Optional[Dict[str, str]]:
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: checkpoint conversion mapping not needed here

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 3]
```diff
@@ -78,13 +97,22 @@ def _build_model(self, device: DeviceLikeType) -> nn.Module:
...
+        return getattr(self, "_checkpoint_conversion_mapping", None)
+
+    def get_exclude_modules_conversion_mapping(self) -> Optional[Dict[str, str]]:
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: unnecessary getter

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 4]
```diff
@@ -45,6 +46,24 @@
 from .hf import AutoModelForCausalLMFactory
 
 
+def _apply_conversion_mapping(pattern: str, conversion_mapping: Dict[str, str]) -> str:
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: don't need a separate function for this, just do it in the remap function

File: tensorrt_llm/_torch/auto_deploy/models/eagle.py [hunk 5]
```diff
@@ -45,6 +46,24 @@
...
+    mapped_pattern = pattern
+    for regex, replacement in conversion_mapping.items():
+        mapped_pattern = re.sub(regex, replacement, mapped_pattern)
+    return mapped_pattern
+
+
+def _remap_exclude_modules_from_conversion_mapping(
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/eagle.py]: maybe just call this "get_renamed_modules()"? This does not need to be so strictly named.

File: tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py [hunk 6]
```diff
@@ -141,6 +141,10 @@ class EagleConfig(PretrainedConfig):
             "_checkpoint_conversion_mapping": {
                 r"^mtp\.": "model.",
             },
+            # Quantization excludes are glob patterns, so handle both mtp.* and mtp*.
+            "_exclude_modules_conversion_mapping": {
```
Comment (govind-ramnarayan) [inline tensorrt_llm/_torch/auto_deploy/models/custom/modeling_eagle.py]: should this name be specifically "_quant_exclude_conversion_mapping"

**Comment by govind-ramnarayan** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1"

File: examples/auto_deploy/model_registry/configs/ultra_v3_mtp.yaml [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        # moe layer: SHARED experts
+        "up_proj": "colwise"
+        "down_proj": "rowwise"
+        # MoLE: latent projections: simple shard
+        "fc1_latent_proj": "gather"
+        "fc2_latent_proj": "gather"
+  multi_stream_moe:
+    enabled: true
+  gather_logits_before_lm_head:
+    # TODO: fix https://github.com/NVIDIA/TensorRT-LLM/issues/9878 to enable by default
```
Comment (nvchenghaoz) [inline examples/auto_deploy/model_registry/configs/ultra_v3_mtp.yaml]: this comment is can be deleted as torch-compile is not a must-have for AD.

**Review by nvchenghaoz — APPROVED** [review]

**Review by jieli-matrix — APPROVED** [review]

**Comment by govind-ramnarayan** [comment]
@coderabbitai Regarding the QA scheduled-list nitpick on `test_mtp`: we add the new `TestNemotronUltraV3::test_mtp` cases in the same places UltraV3 is currently scheduled (the l0 test-db), so no separate `tests/integration/test_lists/qa/` entry is needed for this PR.

**Comment by govind-ramnarayan** [comment]
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> `@govind-ramnarayan` Got it, thanks for the clarification! Since `test_mtp` follows the same scheduling pattern as the existing UltraV3 tests (l0 test-db only), the absence of a separate `tests/integration/test_lists/qa/` entry makes sense. I'll keep this convention in mind for future reviews.
> 
> ✏️ Learnings added
> ```
> Learnt from: govind-ramnarayan
> URL: https://github.com/NVIDIA/TensorRT-LLM/pull/14473
> 
> Timestamp: 2026-05-29T18:49:42.018Z
> Learning: In the TensorRT-LLM repository, new `TestNemotronUltraV3` test cases (e.g., `test_mtp`) do not require a separate `tests/integration/test_lists/qa/` entry. They follow the same scheduling pattern as existing UltraV3 tests and are only added to the l0 test-db (`tests/integration/test_lists/test-db/l0_dgx_b200.yml`).
> ```
> 
> > Note: Learnings are effective only in the context of similar code segments. To apply general coding guidelines, please configure [review instructions](https://docs.coderabbit.ai/guides/review-instructions). You can manage existing learnings in the [app](https://app.coderabbit.ai/learnings).

@coderabbitai Rather, the learning should be that when we add MTP tests in AutoDeploy for a model that is already tested with the base model only, we will just add the tests to the same lists that the base model tests are in, unless explicitly told otherwise.

**Comment by govind-ramnarayan** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by govind-ramnarayan** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by govind-ramnarayan** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"

**Comment by govind-ramnarayan** [comment]
/bot run --stage-list "A10-Build_Docs, A10-PackageSanityCheck-PY310-UB2204, A100X-PackageSanityCheck-PY312-UB2404, A30-AutoDeploy-1, H100_PCIe-AutoDeploy-1, DGX_B200-AutoDeploy-1, A100X-PyTorch-1, DGX_H100-4_GPUs-AutoDeploy-1, DGX_B200-4_GPUs-AutoDeploy-1, DGX_H100-4_GPUs-AutoDeploy-Post-Merge-1, DGX_B200-8_GPUs-AutoDeploy-Post-Merge-1"
